Add sentry integration module

bwrsandman opened the pull request:
Module for enabling raven initializing to report to sentry.

Based on feedback from my presentation at odoo days 2014

Related project: https://github.com/gisce/openerp-sentry
## Todo
- [x] Add git module version overwrite to raven
- [x] Overwrite the except_orm message pop-up with something a little more friendly to the user
- [x] Don't log `SessionExpiredException`
- [x] Don't log `TransactionRollbackError` ?
- [x] Do a better job at sanitizing the passwords (openerp/netsvc.py in dispatch_rpc )
  - Maybe exclude dispatch_rpc from the traceback
- [ ] ~~Enable tests with sentry server in travis~~

Commit:
Add sentry integration module

Module for enabling raven initializing to report to sentry.

Based on feedback from my presentation at odoo days 2014

nbessi commented on sentry/__init__.py:
```diff
@@ -0,0 +1,49 @@
+###############################################################################
+
+from openerp.tools import config
+from odoo_sentry_client import BzrClient
```
from .odoo_sentry_client maybe better

nbessi commented on sentry/__init__.py:
```diff
@@ -0,0 +1,49 @@
+
+from openerp.tools import config
+from odoo_sentry_client import BzrClient
+from odoo_sentry_handler import OdooSentryHandler
```
from .odoo_sentry_handler

nbessi commented on sentry/__init__.py:
```diff
@@ -0,0 +1,49 @@
+    'raven.processors.SanitizePasswordsProcessor',
+    'raven_sanitize_openerp.OpenerpPasswordsProcessor'
+)
+if config.get(u'sentry_dsn'):
```
Is this code executed even if addons is not installed ? if yes it needs fixing

bwrsandman replied:
Sorry I don't understand your objection here

nbessi commented on sentry/__openerp__.py:
```diff
@@ -0,0 +1,62 @@
+Integration with Sentry Error reporting engine.
+
+Insert sentry DSN to ~/.openerp_serverrc with value:
+    sentry_dsn = sync+<Your Sentry DSN>
```
This is a little too short at least information what to see and what to do and what will the addons exactly do.
For what I understand it will redirect Odoo exception to configured Sentry server. I will also add sample of parameter in an Odoo configuration file

nbessi commented on sentry/odoo_sentry_client.py:
```diff
@@ -0,0 +1,104 @@
+import os
+import sys
+from raven import Client
+from bzrlib.branch import Branch
```
Please protect the python import with except log block. It may break server start-up if dependencies are not installed on host.

legalsylvain replied:
Is it necessary to try / except with that statement in __ openerp __ file ?

```
'external_dependencies': {
'python': ['raven', 'raven_sanitize_openerp', 'bzrlib'],
},
```

bwrsandman replied:
The reasoning behind that was to never have sentry cause exceptions.

In the event that you have the module installed, but for some reason or another, one of the python dependencies disappears or causes an error, we don't want sentry itself to start causing crashes.

legalsylvain replied:
How, I think that it is necessary to change the travis file project, to install extra lib, that could not be present on Travis System.
Like in my PR: https://github.com/OCA/server-tools/pull/50/files#diff-0

This should fix Travis: https://travis-ci.org/OCA/server-tools/jobs/33630905#L310

Don't you think ?
Regards.

nbessi commented on sentry/odoo_sentry_client.py:
```diff
@@ -0,0 +1,104 @@
+    versions"""
+
+    def __init__(self, dsn=None, **options):
+        # Send the following library versions, include all eggs
```
I did not understand the docstring

nbessi commented on sentry/odoo_sentry_client.py:
```diff
@@ -0,0 +1,104 @@
+
+    def set_rev_version(self, path):
+        """Given path, get source and revno, careful not to raise any
+        exceptions"""
```
I would add because it is used in error management

nbessi commented:
Nice addition. It would be nice to add a command line argument too.
Some comment in line

bwrsandman commented:
Thanks for the review @nbessi, I will get to it eventually :tired_face:

nbessi commented:
@bwrsandman I did not dare to talk about unit test :sweat:

gurneyalex commented:
tests are red because of missing external dependency, unhandled in the imports. This must be taken care of because the sole presence of the addon in the repository without the `sentry` python module will cause a crash at instance startup.

nbessi commented:
@gurneyalex this was allready pointed as a need fixing point in sentry/odoo_sentry_client.py line 27,28

bwrsandman commented:
@nbessi, @gurneyalex, Idealy, I would like to be able to overwrite the error message displayed by the client.

Instead of displaying a traceback to the user, it would be a lot better to show a message saying that an error occured and was sent to sentry.

Do you have an idea how I can do this?

legalsylvain commented:
@bwrsandman, maybe that feature should be integrated to another module like:
'web_display_friendly_message' in https://github.com/OCA/web
It' not just a need of your module I think.
What do you think ?
I could work on that if you want but in a while.

Do you need help for other think ? I'm very interested by your module and Sentry system, so if I can help for something.

Commit:
Overwrite web error message to hide traceback from user

Inject sentry_activated into web session
Overwrite CrashManager.error QWeb template
Provide translations

Signed-off-by: Sandy Carter <sandy.carter@savoirfairelinux.com>

Commit:
Skip Uncaught exception in osv which is a duplicate error

legalsylvain commented:
Hi @bwrsandman, 
Thanks for all the changes. (Friendly messages to users, ...) 
I planned to review this PR with @arthru, realizing complete tests with sentry service installed in one or two weeks.
I'll let you know.
Regards.

kryskool commented on sentry_logger/__openerp__.py:
```diff
@@ -0,0 +1,66 @@
+    # any module necessary for this one to work correctly
+    'depends': ['web'],
+    'external_dependencies': {
+        'python': ['raven', 'raven_sanitize_openerp', 'bzrlib', 'git'],
```
Hi

Please removed "bzrlib" and "git" dependancies on **openerp**.py, because i only use git, and convert bzr repo with fast-export and fast-import, and i cannot install this module. The protection on the odoo_sentry_client.py file is adequacy to check if bzrlib and git is installed

Regards,

bwrsandman replied:
The code should work with and without git, bzr python libraries, so I see nothing wrong with having these be optional dependencies.

bwrsandman replied:
I set bzrlib as optional, but let git in, since this repo is git and we would like to always have sentry_logger report the version of itself.

Commit:
Set bzrlib as optional dependency.

Code should work without bzrlib and skip logging the revnos if lib is absent.

Git dependency kept since oca/server-tools itself is git and we would like to
know sha of sentry_logger

Commit:
Don't log except_osv exceptions

Commit:
Don't log SessionExpiredException and TransactionRollbackError

Commit:
Exclude dispatch and dispatch_rpc frames as they contain passwords

bwrsandman commented:
Raven no longer logs passwords. Note that I had trouble with python-raven 5.0.0, using 5.1.1 seems to work.

legalsylvain commented:
Could you rebase ? Travis fails for fixed pylint errors.

Regards.

bwrsandman commented:
@legalsylvain not rebased, but tests restarted

legalsylvain commented:
- Code Reviewed.
- Tests realized with sentry server.
  :+1: 
  All is OK. Thanks a lot @bwrsandman.

arthru commented:
:+1:

Reviewed, tested, works like a charm !

bwrsandman commented:
Small note and something to test: it impacts performance of spawning workers.

Since each worker fetches the config, it can take up to a second more for workers to spawn.

legalsylvain commented:
Can you add this note in the openerp file ?

Commit:
Add warning about performance impact

bwrsandman commented:
Added

pedrobaeza commented:
:+1:

bwrsandman commented:
Thank you for the review, @legalsylvain  and @arthru. I hope this module serves you as well as it has for me!
